File: metadata/mill-metadata-core/src/main/java/io/qpointz/mill/metadata/service/FacetCatalog.java
package io.qpointz.mill.metadata.service;

import io.qpointz.mill.metadata.domain.FacetTypeDescriptor;
import io.qpointz.mill.metadata.domain.MetadataEntity;
import io.qpointz.mill.metadata.domain.MetadataTargetType;
import io.qpointz.mill.metadata.domain.ValidationResult;

import java.util.Collection;
import java.util.Optional;

public interface FacetCatalog {

    void register(FacetTypeDescriptor descriptor);
    void update(FacetTypeDescriptor descriptor);
    void delete(String typeKey);

    Optional<FacetTypeDescriptor> get(String typeKey);
    Collection<FacetTypeDescriptor> getAll();
    Collection<FacetTypeDescriptor> getEnabled();
    Collection<FacetTypeDescriptor> getMandatory();
    Collection<FacetTypeDescriptor> getForTargetType(MetadataTargetType targetType);

    boolean isAllowed(String typeKey);
    boolean isMandatory(String typeKey);
    boolean isApplicableTo(String typeKey, MetadataTargetType targetType);

    ValidationResult validateFacetContent(String typeKey, Object facetData);
    ValidationResult validateEntityFacets(MetadataEntity entity);
}

File: metadata/mill-metadata-core/src/main/kotlin/io/qpointz/mill/metadata/service/FacetCatalog.kt
package io.qpointz.mill.metadata.service

import io.qpointz.mill.metadata.domain.FacetType
import io.qpointz.mill.metadata.domain.FacetTypeDefinition
import io.qpointz.mill.metadata.domain.ValidationResult
import io.qpointz.mill.metadata.domain.facet.FacetTargetCardinality

/** Facet type catalog (SPEC §7.3). */
interface FacetCatalog {
    fun findType(typeKey: String): FacetType?
    fun findDefinition(typeKey: String): FacetTypeDefinition?
    fun listDefinitions(): List<FacetTypeDefinition>
    fun listTypes(): List<FacetType>
    fun inspect(typeKey: String, payload: Map<String, Any?>): ValidationResult
    fun resolveCardinality(typeKey: String): FacetTargetCardinality
    fun registerDefinition(definition: FacetTypeDefinition): FacetTypeDefinition
}
